File: src/main/java/BSCLM/Controller.java
package BSCLM;

import javafx.collections.FXCollections;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.Pane;
import javafx.stage.DirectoryChooser;

import java.io.File;
import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.Optional;

public class Controller {

    @FXML
    private ListView<Song> song_list;
    @FXML
    private Pane pane;
    @FXML
    private ImageView song_image;
    @FXML
    private Label song_name;
    @FXML
    private Label song_sub_name;
    @FXML
    private Label song_author_name;
    @FXML
    private Label level_author_name;

    private ArrayList<Song> song;
    private File folder;

    final OGGPlayer player = new OGGPlayer();

    @FXML
    public void deleteSong()
    {
        Alert alert = new Alert(Alert.AlertType.CONFIRMATION);
        alert.setTitle("删除确认");
        String s ="你将要删除 " + song_list.getSelectionModel().getSelectedIndices().size() + " 项文件";
        alert.setContentText(s);
        Optional<ButtonType> result = alert.showAndWait();

        if ((result.isPresent()) && (result.get() == ButtonType.OK)) {
            int last_index = -1;
            player.stop();
            for (int i: song_list.getSelectionModel().getSelectedIndices().sorted((o1, o2) -> o2-o1)){
                System.out.print(deleteFolder(song.get(i).getPath()));
                song.remove(i);
                last_index = i;
            }

            song_list.setItems(FXCollections.observableArrayList(song));
            song_list.getSelectionModel().select(last_index);
            song_list.requestFocus();
        }
    }

    @FXML
    public void clearAllSong()
    {
        song_list.getSelectionModel().clearSelection();
    }

    @FXML
    public void selectFolder()
    {
        DirectoryChooser folderChooser = new DirectoryChooser();
        folderChooser.setTitle("选择BeatSaber自定义曲目文件夹");
        folder = folderChooser.showDialog(pane.getScene().getWindow());
        if (folder != null)
            initialize();
    }

    @FXML
    private void initialize()
    {
        song = new ArrayList<>();
        if(folder == null) return;

        File[] files = folder.listFiles();

        if (files == null) return;

        for (File file: files) {
            if (!file.isDirectory()) continue;
            try {
                Song s = new Song(file);
                song.add(s);
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        }

        song.sort(Comparator.comparing(Song::getSongName));
        song_list.setCellFactory(new SongCellFactory());
        song_list.getSelectionModel().selectedItemProperty().addListener(
                (observable, oldValue, newValue) -> {
                    if (newValue == null) return;
                        song_image.setImage(new Image(new File(newValue.getSongImage()).toURI().toString()));
                        song_name.setText(newValue.getSongName());
                        song_sub_name.setText(newValue.getSongSubName());
                        song_author_name.setText(newValue.getSongAuthorName());
                        level_author_name.setText(newValue.getLevelAuthorName());
                        player.play(newValue.getSong());
                });

        song_list.setItems(FXCollections.observableArrayList(song));
        song_list.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);

    }

    private boolean deleteFolder(File folder) {
        if (!folder.exists()) return false;
        if (folder.isDirectory()) {
            String[] children = folder.list();
            if (children != null)
                for (String child : children) {
                    boolean bool = deleteFolder(new File(folder, child));
                    if (!bool) return false;
                }
        }
        return folder.delete();
    }
}

File: src/main/java/BSCLM/Controller.kt
package BSCLM

import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.layout.Pane
import javafx.stage.DirectoryChooser

import java.io.File
import java.util.*

class Controller {

    @FXML
    private lateinit var songList: ListView<Song>
    @FXML
    private lateinit var pane: Pane
    @FXML
    private lateinit var songImage:ImageView
    @FXML
    private lateinit var songName:Label
    @FXML
    private lateinit var songSubName:Label
    @FXML
    private lateinit var songAuthorName: Label
    @FXML
    private lateinit var levelAuthorName: Label
    @FXML
    private lateinit var btnDelete: Button
    @FXML
    private lateinit var btnClear: Button

    private var song = ArrayList<Song>()
    private var folder: File? = null
    private val res = ResourceBundle.getBundle("string")
    private val player = OGGPlayer()

    @FXML
    fun deleteSong() {
        val result = Alert(Alert.AlertType.CONFIRMATION).apply {
            title = res.getString("delete.confirm.title")
            contentText = "${res.getString("delete.confirm.msg.first")} ${songList.selectionModel.selectedIndices.count()} ${res.getString("delete.confirm.msg.last")}"
        }.showAndWait()

        if ((result.isPresent) && (result.get() == ButtonType.OK)) {
            var last_index = -1
            player.stop()
            with(songList){
                for (i in selectionModel.selectedIndices.sorted()){
                    print(deleteFolder(File(song[i].getPath())))
                    song.removeAt(i)
                    last_index = i
                }

                items = FXCollections.observableArrayList(song)
                selectionModel.select(last_index)
                requestFocus()
            }
        }
    }

    @FXML
    fun clearAllSong() {
        songList.selectionModel.clearSelection()
    }

    @FXML
    fun selectFolder() {
        val folderChooser = DirectoryChooser()
        folderChooser.title = res.getString("folderChooser.title")
        folder = folderChooser.showDialog(pane.scene.window)
        if (folder != null)
            initialize()
    }

    @FXML
    fun initialize() {
        song.clear()
        folder?:return

        val files = folder!!.listFiles() ?: return
        var successCount = 0
        for (file in files) {
            if (!file.isDirectory) continue
            try {
                val s = Song(file.path)
                song.add(s)
                successCount ++
            } catch (e: Exception) {
                print(e.message)
            }
        }

        Alert(Alert.AlertType.INFORMATION).apply {
            title = res.getString("import.title")
            contentText = "${res.getString("import.msg.first")} $successCount/${files.count()} ${res.getString("import.msg.last")}"
        }.showAndWait()

        song.sortWith(Comparator.comparing(Song::getSongName))
        with(songList) {
            cellFactory = SongCellFactory()
            selectionModel.selectedItemProperty().addListener { _, _, newValue ->
                newValue?: run { btnDelete.isDisable = true; btnClear.isDisable = true; return@addListener }
                songImage.image = Image(File(newValue.getSongImage()).toURI().toString())
                songName.text = newValue.getSongName()
                songSubName.text = newValue.getSongSubName()
                songAuthorName.text = newValue.getSongAuthorName()
                levelAuthorName.text = newValue.getLevelAuthorName()
                player.play(newValue.getSong())
                System.gc()
                btnDelete.isDisable = false
                btnClear.isDisable = false
            }
            items = FXCollections.observableArrayList(song)
            selectionModel.selectionMode = SelectionMode.MULTIPLE
        }
    }

    private fun deleteFolder(folder: File):Boolean {
        if (!folder.exists()) return false
        if (folder.isDirectory) {
            val children = folder.list()
            if (children != null)
                for (child in children) {
                val bool = deleteFolder(File(folder, child))
                if (!bool) return false
            }
        }
        return folder.delete()
    }
}
